Issues/1445 synapse catalogue widget

clbarnes opened the pull request:
This is the first iteration of the Connector Viewer widget, referred to as the Synapse Catalogue in issue #1445. Users add a number of skeletons to the widget and can peruse the connectors associated with those skeletons in a grid of mini stack viewers (with slightly altered stackInfo bars). They can filter by the type of association, as in the Connector Table. There are a number of sorting algorithms available, and the user can define which 'main' stack viewer they want their mini viewers to inherit settings from. 

Existing issues:

- there's no button associated with the widget as yet
- tracing overlay keyboard shortcuts behave inconsistently in the mini stack viewers
- connector sorting may hit performance issues for large numbers of connectors (untested)
- tracing with the connector viewer open is slowed down quite a lot - this is probably just related to having a lot of stack viewers, so it may be beneficial to suspend the tracing overlay of any stack viewer which is not currently in focus if it's not navigating with the project.

Future enhancements:

- At the moment I'm just jQuerying the stackInfo bar to make it do what I want - it might be nice to be able to customise the stackviewer, possibly by subclassing it, to handle that
- To me, this is a natural partner with the connector table: I experimented with a button to open a connector table from the connector viewer widget but then the table's own skeleton source controls misbehaved. What would make the most sense to me is having them joined like the 3D viewer is with its selection table, using the same sorting algorithm and possibily pagination.

Really really minor thing:

- You can click on the connector name in the stackInfo bar to navigate your main stack viewer to that connector, but that wouldn't be obvious to a user - I'd like to put a little target or drop pin icon next to it but wasn't sure about how to find static files and how not to accidentally use a copyrighted svg and such...

Contributors: @aschampion 

![connectorviewerfinal](https://cloud.githubusercontent.com/assets/8995328/21872005/dde8d8e4-d833-11e6-9bfa-1b0b4bc57cf3.png)

clbarnes commented on django/applications/catmaid/static/js/stack-viewer.js:
```diff
@@ -19,7 +19,8 @@
       project,          //!< {CATMAID.Project} reference to the parent project
-      primaryStack
+      primaryStack,
+      window
```
Oops, I think `window` is reserved, isn't it?

tomka replied:
Yep, it is. Maybe use `catmaidWindow`?

tomka commented:
Looks really nice, the tool itself and code-wise! I try to go through it at some point today. By the way, JSHint complains about some missing semicolons (see [Travis CI](https://travis-ci.org/catmaid/CATMAID/builds/191161187#L703)).

Being able to open a Connector Table would certainly be useful. And if you can load arbitrary sets of connectors (i.e. not only all per skeleton), being able to open a pre-populated viewer from a Connector Table would be nice as well. And I could also see a button in the Connectivity Widget to open a Connector Viewer to be useful, to quickly inspect connector sites visually.

aschampion commented:
Regarding slowness with multiple tracing layers open:

- Right now 3 node refreshes per connector viewer panel are happening for every tracing action. This is also the case for a second stack viewer in general, so is not an issue specific to this widget but one which would also slow down multi view tracing.

- One big CATMAID-wide bottleneck is that all requests end up in the same serialized queue (see also #1162). Instead, the submitter/request queues should be smart enough to distinguish mutating and non-mutating requests, allow sequences of non-mutating requests to be dispatched in parallel, but serialize when a mutating requests hits the top of the queue. It could still invoke callbacks in insertion order. This would preserve current guarantees (except a few cases around queue replacement) without having to enumerate all potential interactions between requests.

But for now I think the performance is fine, and as you suggested just suspending when connector panels when unfocused is a simple and effective solution in the near term.

aschampion commented:
Some errors, probably from stack viewer bookkeeping:

- When using WebGL tile layers, get a bunch of shader errors when opening the connector viewer:

```
ERROR: 0:5: '' : array size must be greater than zero 
 
pixi.min.js:2 WebGL: INVALID_VALUE: attachShader: no object or object deletedi @ pixi.min.js:2a @ pixi.min.js:2o @ pixi.min.js:3i @ pixi.min.js:7i.flush @ pixi.min.js:7i.render @ pixi.min.js:5PixiContext.renderIfReady @ pixi-layer.js:65PixiLayer._renderIfReady @ pixi-layer.js:337PixiTileLayer.redraw @ pixi-tile-layer.js:234TileLayer.resize @ tile-layer.js:414PixiTileLayer.resize @ pixi-tile-layer.js:251(anonymous function) @ stack-viewer.js:562StackViewer.resize @ stack-viewer.js:561ConnectorViewer.redrawPanels @ connector-viewer.js:497ConnectorViewer.init @ connector-viewer.js:398init @ connector-viewer.js:384createWidget @ WindowMaker.js:180creators.(anonymous function) @ WindowMaker.js:2625create @ WindowMaker.js:2580OpenWidgetDialog.dialog.onOK @ open-widget-dialog.js:27OK @ options-dialog.js:44s.click @ jquery-ui.min.js:11dispatch @ jquery-2.1.3.js:4430r.handle @ jquery-2.1.3.js:4116trigger @ jquery-2.1.3.js:4345(anonymous function) @ jquery-2.1.3.js:4896each @ jquery-2.1.3.js:374each @ jquery-2.1.3.js:139trigger @ jquery-2.1.3.js:4895n.fn.(anonymous function) @ jquery-2.1.3.js:7482(anonymous function) @ options-dialog.js:113dispatch @ jquery-2.1.3.js:4430r.handle @ jquery-2.1.3.js:4116
pixi.min.js:2 Pixi.js Error: Could not initialize shader.i @ pixi.min.js:2a @ pixi.min.js:2o @ pixi.min.js:3i @ pixi.min.js:7i.flush @ pixi.min.js:7i.render @ pixi.min.js:5PixiContext.renderIfReady @ pixi-layer.js:65PixiLayer._renderIfReady @ pixi-layer.js:337PixiTileLayer.redraw @ pixi-tile-layer.js:234TileLayer.resize @ tile-layer.js:414PixiTileLayer.resize @ pixi-tile-layer.js:251(anonymous function) @ stack-viewer.js:562StackViewer.resize @ stack-viewer.js:561ConnectorViewer.redrawPanels @ connector-viewer.js:497ConnectorViewer.init @ connector-viewer.js:398init @ connector-viewer.js:384createWidget @ WindowMaker.js:180creators.(anonymous function) @ WindowMaker.js:2625create @ WindowMaker.js:2580OpenWidgetDialog.dialog.onOK @ open-widget-dialog.js:27OK @ options-dialog.js:44s.click @ jquery-ui.min.js:11dispatch @ jquery-2.1.3.js:4430r.handle @ jquery-2.1.3.js:4116trigger @ jquery-2.1.3.js:4345(anonymous function) @ jquery-2.1.3.js:4896each @ jquery-2.1.3.js:374each @ jquery-2.1.3.js:139trigger @ jquery-2.1.3.js:4895n.fn.(anonymous function) @ jquery-2.1.3.js:7482(anonymous function) @ options-dialog.js:113dispatch @ jquery-2.1.3.js:4430r.handle @ jquery-2.1.3.js:4116
pixi.min.js:2 gl.VALIDATE_STATUS falsei @ pixi.min.js:2a @ pixi.min.js:2o @ pixi.min.js:3i @ pixi.min.js:7i.flush @ pixi.min.js:7i.render @ pixi.min.js:5PixiContext.renderIfReady @ pixi-layer.js:65PixiLayer._renderIfReady @ pixi-layer.js:337PixiTileLayer.redraw @ pixi-tile-layer.js:234TileLayer.resize @ tile-layer.js:414PixiTileLayer.resize @ pixi-tile-layer.js:251(anonymous function) @ stack-viewer.js:562StackViewer.resize @ stack-viewer.js:561ConnectorViewer.redrawPanels @ connector-viewer.js:497ConnectorViewer.init @ connector-viewer.js:398init @ connector-viewer.js:384createWidget @ WindowMaker.js:180creators.(anonymous function) @ WindowMaker.js:2625create @ WindowMaker.js:2580OpenWidgetDialog.dialog.onOK @ open-widget-dialog.js:27OK @ options-dialog.js:44s.click @ jquery-ui.min.js:11dispatch @ jquery-2.1.3.js:4430r.handle @ jquery-2.1.3.js:4116trigger @ jquery-2.1.3.js:4345(anonymous function) @ jquery-2.1.3.js:4896each @ jquery-2.1.3.js:374each @ jquery-2.1.3.js:139trigger @ jquery-2.1.3.js:4895n.fn.(anonymous function) @ jquery-2.1.3.js:7482(anonymous function) @ options-dialog.js:113dispatch @ jquery-2.1.3.js:4430r.handle @ jquery-2.1.3.js:4116
pixi.min.js:2 gl.getError() 1281i @ pixi.min.js:2a @ pixi.min.js:2o @ pixi.min.js:3i @ pixi.min.js:7i.flush @ pixi.min.js:7i.render @ pixi.min.js:5PixiContext.renderIfReady @ pixi-layer.js:65PixiLayer._renderIfReady @ pixi-layer.js:337PixiTileLayer.redraw @ pixi-tile-layer.js:234TileLayer.resize @ tile-layer.js:414PixiTileLayer.resize @ pixi-tile-layer.js:251(anonymous function) @ stack-viewer.js:562StackViewer.resize @ stack-viewer.js:561ConnectorViewer.redrawPanels @ connector-viewer.js:497ConnectorViewer.init @ connector-viewer.js:398init @ connector-viewer.js:384createWidget @ WindowMaker.js:180creators.(anonymous function) @ WindowMaker.js:2625create @ WindowMaker.js:2580OpenWidgetDialog.dialog.onOK @ open-widget-dialog.js:27OK @ options-dialog.js:44s.click @ jquery-ui.min.js:11dispatch @ jquery-2.1.3.js:4430r.handle @ jquery-2.1.3.js:4116trigger @ jquery-2.1.3.js:4345(anonymous function) @ jquery-2.1.3.js:4896each @ jquery-2.1.3.js:374each @ jquery-2.1.3.js:139trigger @ jquery-2.1.3.js:4895n.fn.(anonymous function) @ jquery-2.1.3.js:7482(anonymous function) @ options-dialog.js:113dispatch @ jquery-2.1.3.js:4430r.handle @ jquery-2.1.3.js:4116
pixi.min.js:2 Pixi.js Warning: gl.getProgramInfoLog() missing shaders i @ pixi.min.js:2a @ pixi.min.js:2o @ pixi.min.js:3i @ pixi.min.js:7i.flush @ pixi.min.js:7i.render @ pixi.min.js:5PixiContext.renderIfReady @ pixi-layer.js:65PixiLayer._renderIfReady @ pixi-layer.js:337PixiTileLayer.redraw @ pixi-tile-layer.js:234TileLayer.resize @ tile-layer.js:414PixiTileLayer.resize @ pixi-tile-layer.js:251(anonymous function) @ stack-viewer.js:562StackViewer.resize @ stack-viewer.js:561ConnectorViewer.redrawPanels @ connector-viewer.js:497ConnectorViewer.init @ connector-viewer.js:398init @ connector-viewer.js:384createWidget @ WindowMaker.js:180creators.(anonymous function) @ WindowMaker.js:2625create @ WindowMaker.js:2580OpenWidgetDialog.dialog.onOK @ open-widget-dialog.js:27OK @ options-dialog.js:44s.click @ jquery-ui.min.js:11dispatch @ jquery-2.1.3.js:4430r.handle @ jquery-2.1.3.js:4116trigger @ jquery-2.1.3.js:4345(anonymous function) @ jquery-2.1.3.js:4896each @ jquery-2.1.3.js:374each @ jquery-2.1.3.js:139trigger @ jquery-2.1.3.js:4895n.fn.(anonymous function) @ jquery-2.1.3.js:7482(anonymous function) @ options-dialog.js:113dispatch @ jquery-2.1.3.js:4430r.handle @ jquery-2.1.3.js:4116
pixi.min.js:2 WebGL: INVALID_VALUE: getProgramParameter: no object or object deleteds @ pixi.min.js:2a @ pixi.min.js:2o @ pixi.min.js:3i @ pixi.min.js:7i.flush @ pixi.min.js:7i.render @ pixi.min.js:5PixiContext.renderIfReady @ pixi-layer.js:65PixiLayer._renderIfReady @ pixi-layer.js:337PixiTileLayer.redraw @ pixi-tile-layer.js:234TileLayer.resize @ tile-layer.js:414PixiTileLayer.resize @ pixi-tile-layer.js:251(anonymous function) @ stack-viewer.js:562StackViewer.resize @ stack-viewer.js:561ConnectorViewer.redrawPanels @ connector-viewer.js:497ConnectorViewer.init @ connector-viewer.js:398init @ connector-viewer.js:384createWidget @ WindowMaker.js:180creators.(anonymous function) @ WindowMaker.js:2625create @ WindowMaker.js:2580OpenWidgetDialog.dialog.onOK @ open-widget-dialog.js:27OK @ options-dialog.js:44s.click @ jquery-ui.min.js:11dispatch @ jquery-2.1.3.js:4430r.handle @ jquery-2.1.3.js:4116trigger @ jquery-2.1.3.js:4345(anonymous function) @ jquery-2.1.3.js:4896each @ jquery-2.1.3.js:374each @ jquery-2.1.3.js:139trigger @ jquery-2.1.3.js:4895n.fn.(anonymous function) @ jquery-2.1.3.js:7482(anonymous function) @ options-dialog.js:113dispatch @ jquery-2.1.3.js:4430r.handle @ jquery-2.1.3.js:4116
pixi.min.js:2 WebGL: INVALID_VALUE: getProgramParameter: no object or object deleteds @ pixi.min.js:2a @ pixi.min.js:2o @ pixi.min.js:3i @ pixi.min.js:7i.flush @ pixi.min.js:7i.render @ pixi.min.js:5PixiContext.renderIfReady @ pixi-layer.js:65PixiLayer._renderIfReady @ pixi-layer.js:337PixiTileLayer.redraw @ pixi-tile-layer.js:234TileLayer.resize @ tile-layer.js:414PixiTileLayer.resize @ pixi-tile-layer.js:251(anonymous function) @ stack-viewer.js:562StackViewer.resize @ stack-viewer.js:561ConnectorViewer.redrawPanels @ connector-viewer.js:497ConnectorViewer.init @ connector-viewer.js:398init @ connector-viewer.js:384createWidget @ WindowMaker.js:180creators.(anonymous function) @ WindowMaker.js:2625create @ WindowMaker.js:2580OpenWidgetDialog.dialog.onOK @ open-widget-dialog.js:27OK @ options-dialog.js:44s.click @ jquery-ui.min.js:11dispatch @ jquery-2.1.3.js:4430r.handle @ jquery-2.1.3.js:4116trigger @ jquery-2.1.3.js:4345(anonymous function) @ jquery-2.1.3.js:4896each @ jquery-2.1.3.js:374each @ jquery-2.1.3.js:139trigger @ jquery-2.1.3.js:4895n.fn.(anonymous function) @ jquery-2.1.3.js:7482(anonymous function) @ options-dialog.js:113dispatch @ jquery-2.1.3.js:4430r.handle @ jquery-2.1.3.js:4116
pixi.min.js:7 WebGL: INVALID_OPERATION: drawElements: no valid shader program in use
```

- Exception:
  1. Open the connector viewer
  2. Close it
  3. Open another stack viewer
  4. Will get an exception on `overlay.js:2159` because `this.submit` is `null`. Sometimes I don't get the exception here, but instead when closing the newly opened stack viewer, because a mouse catcher is null.

clbarnes commented:
Thanks! I'll have a look into these.

Something which I think needs checking is that I may have gotten carried away with function binding - I suspect it's not necessary in a few places, particularly in the `onclick` and `onchange` events in the widget controls.

@tomka  - at the moment the whole caching and ordering system works based on the underlying skeleton source so opening arbitrary connectors would be tricky, although I agree that looking specifically at the connectors between two skeletons (or from one specified set to another specified set of skeletons), for example as loaded from the connectivity widget, could be useful. The more I think about this sort of thing, the more I want to do away with skeleton source controls of the connector viewer entirely and make it a direct slave to the connector table (which could do all the filtering itself), but it would mean moving all of the sorting, caching etc. code over to the latter and adding a bunch of new features to it. Probably out of scope right now. I have a 'Table' button on the Viewer working right now but bizarrely, the exact same code can't open a Viewer from the Table; it complains that particular skeletons don't exist. I'll look into this.

clbarnes commented:
> I have a 'Table' button on the Viewer working right now but bizarrely, the exact same code can't open a Viewer from the Table; it complains that particular skeletons don't exist. I'll look into this.

There was a bug in the connector table where the result skeleton source wasn't populated properly, fixed now.

clbarnes commented:
These builds are failing because download.slimerjs.org is down, it seems.

aschampion commented:
Regarding 3 node refreshes on unfocused tracing overlays per tracing action: these are due to subscriptions for multiple events (new node, skeleton changed, etc.) being triggered and invoking `updateNodes`. Node query requests are supposed to be deduplicated/pre-empted by request replacement in the submitter, but the submitter is not actually replacing requests because it doesn't replace anything until it's left the submitter queue and placed in the request queue. Of course nothing is popped from the submitter queue until it's already been handled by the request queue, thus no replacement. Also the cause of #1127.

The story for queue replacement has always been a bit vague, and especially so now with promises and strong assumptions about prior requests succeeding.

clbarnes commented:
Remaining issues:

- WebGL issues as per @aschampion's post
- Whenever a node is placed in a mini stack viewer, it returns focus to the widget (i.e. you need to click again before the mini stack viewer regains focus and you can place another node) 
- Performance issues are worked around by suspending every other mini stack viewer when one gains focus, and forcing an update when they gain focus again
- slimerJS.org is rubbish

aschampion commented:
SlimerJS.org...has been dealt with :scissors: :wave: :skull: e917e57f8ce6ec2f8f7f73e4a2084d393745e016

tomka commented:
> function binding - I suspect it's not necessary in a few places, particularly in the onclick and onchange events in the widget controls.

True. Instead of writing `self.updateWithSkels.bind(self)();` you can just write `self.updateWithSkels();`. This will will avoid the creation of an extra closure through `bind()`, provides the context for the function directly and is more readable.

> The more I think about this sort of thing, the more I want to do away with skeleton source controls of the connector viewer entirely and make it a direct slave to the connector table

Merging the two sounds reasonable, but is nothing that needs to be done right away.

It would be nice if you could add a few sentences to the `CHANGELOG.md` file. And if it possible to summarize basic usage you could provide it as text for the `helpText` field in returned object of `getWidgetConfiguration()`.

clbarnes commented:
Changelog, help text, and a few bug fixes in a push going up now.

Remaining issues:

- WebGL issues as per @aschampion's post
- Whenever a node is placed in a mini stack viewer, it returns focus to the widget (i.e. you need to click again before the mini stack viewer regains focus and you can place another node)
- Adding buttons

aschampion commented:
After debugging with @clbarnes the WebGL error is a Pixi error that's fixed by V4.2.2 or earlier. Since it's not breaking, it doesn't need to be fixed in this PR. I'll update Pixi to a safe fixed version separately in `dev`.

tomka commented on django/applications/catmaid/static/js/widgets/connector-viewer.js:
```diff
@@ -0,0 +1,1118 @@
+      this.update();
+      return Number(currentPageElement.value) - 1;
+    } else if (newPage < 0 || newFirstConnectorIdx >= total) {  // page out of bounds
+      alert('This page does not exist! Returning to page 1.');
```
Instead of `alert()`, I think `CATMAID.error()` would be nicer, because most errors are reported that way. However, maybe `CATMAID.warn()` would be enough to not interrupt too much due a wrong page number.

tomka commented:
Just tested it out locally and it works really well. Sure, there are still some update performance issues, but these should be sorted out separately (like we already discussed). Great that the WebGL problems can be fixed with a Pixi.js update.

I also went through the code, but couldn't see any obvious problem. And since we we don't really support IE, using template literals for strings and non-basic Set features seems okay to me.

If there are no objections, I am fine with merging this.

clbarnes commented:
Fortunately, @ceesem came by and made me demo it, and as is the law of software demos, it broke when the grid was resized but that should be a one-line change.

I'm still trying to sort out why focus switches back to the widget window every time you place a node; I'd like that to be fixed before the merge. And I'm open to suggestions on how to make the widget accessible - it can currently be opened with ctrl-space or through the connector table, but I think the connector table is pretty obscure too so it may need a button.

tomka commented:
Hm, grid resizing worked for me, but it's a good thing if you can making it more robust. Also I noticed that one can enter letters into the page input field and the out-of-range check doesn't handle `NaN`s, so no error/warning is shown. And having the focus fixed would be nice, I agree. In any case, let us know if you think it is ready.

clbarnes commented:
I'm happy to call this done, if you don't mind checking out the last couple of changes - mainly I refactored the sorting. I rebased on dev this morning so it should be ready to go.

Something which should have an easy fix (but I can't figure it out) is the focus pinging back to the widget whenever you place a node in a mini stack viewer. 

Larger enhancements which could be spun out into their own issues:

- Sort by connector creator (may require an API change)
- Constrain connector selection by 'source skeletons' and 'target skeletons' (rather than having a single skeleton selection)
- Unify connector-viewer and connector-table (and maybe create-connector-selection?) - so they could share skeleton sources, sorting information and so on

tomka commented:
Nice! I read through the code and tested it locally, it all seems to work and I am happy to merge it. Let's figure out the focus issue separately. The three enhancements you list sound all reasonable and should get each a new Github issue.
